crypto/cmp/cmp_vfy.c

```diff
@@ -0,0 +1,868 @@
+        }
+    } else {
+        BIO_printf(bio, "    (no certificates)\n");
+    }
+}
+
+static void print_store_certs(BIO *bio, X509_STORE *store)
+{
+    if (store != NULL) {
+        STACK_OF(X509) *certs = ossl_cmp_X509_STORE_get1_certs(store);
+        print_certs(bio, certs);
+        sk_X509_pop_free(certs, X509_free);
+    } else {
+        BIO_printf(bio, "    (no certificate store)\n");
+    }
+}
+
+/*
+ * Diagnostic function that may be registered using
+ * X509_STORE_set_verify_cb(), such that it gets called by OpenSSL's
+ * verify_cert() function at the end of a cert verification as an opportunity
+ * to gather and queue information regarding a (failing) cert verification,
+ * and to possibly change the result of the verification (not done here).
+ * The CLI also calls it on error while cert status checking using OCSP stapling
+ * via a callback function set with SSL_CTX_set_tlsext_status_cb().
+ * returns 0 if and only if the cert verification is considered failed.
+ */
+int OSSL_CMP_print_cert_verify_cb(int ok, X509_STORE_CTX *ctx)
+{
+    if (ok == 0 && ctx != NULL) {
+        int cert_error = X509_STORE_CTX_get_error(ctx);
+        int depth = X509_STORE_CTX_get_error_depth(ctx);
+        X509 *cert = X509_STORE_CTX_get_current_cert(ctx);
+        BIO *bio = BIO_new(BIO_s_mem()); /* may be NULL */
+
+        if (depth < 0)
+            BIO_printf(bio, "signature verification ");
+        else
+            BIO_printf(bio, "%s at depth=%d error=%d (%s)\n",
+                       X509_STORE_CTX_get0_parent_ctx(ctx) != NULL
+                       ? "CRL path validation" : "certificate verification",
+                       depth, cert_error,
+                       X509_verify_cert_error_string(cert_error));
+        BIO_printf(bio, "failure for:\n");
+        print_cert(bio, cert, X509_FLAG_NO_EXTENSIONS);
+        if (cert_error == X509_V_ERR_CERT_UNTRUSTED
+                || cert_error == X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT
+                || cert_error == X509_V_ERR_SELF_SIGNED_CERT_IN_CHAIN
+                || cert_error == X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT
+                || cert_error == X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT_LOCALLY
+                || cert_error == X509_V_ERR_UNABLE_TO_GET_CRL_ISSUER
+                || cert_error == X509_V_ERR_STORE_LOOKUP) {
+            BIO_printf(bio, "non-trusted certs:\n");
+            print_certs(bio, X509_STORE_CTX_get0_untrusted(ctx));
+            BIO_printf(bio, "trust store:\n");
+            print_store_certs(bio, X509_STORE_CTX_get0_store(ctx));
+        }
+        CMPerr(0, CMP_R_POTENTIALLY_INVALID_CERTIFICATE);
+        if (bio != NULL) {
+            char *str;
+            long len = BIO_get_mem_data(bio, &str);
+
+            if (len > 0) {
+                str[len-1] = '\0'; /* replace last '\n', terminating str */
+                ossl_cmp_add_error_line(str);
+            }
+        }
+        BIO_free(bio);
+    }
+#if 0
```

Comment:
I prefer not to add "#if 0" code unless there is a strong justification. Isn't just the TODO comment sufficient?

Reply:
Fixed.